waiter/integration/waiter/token_watch_integration_test.clj

```diff
@@ -0,0 +1,362 @@
+
+(defn- start-watch
+  [router-url cookies & {:keys [query-params] :or {query-params {"include" ["metadata", "deleted"]
+                                                                 "watch" "true"}}}]
+  (let [{:keys [body error]} (make-request router-url "/tokens"
+                                           :cookies cookies
+                                           :fold-chunked-response? false
+                                           :query-params query-params)
+        token->index-atom (atom {})
+        query-state-fn (fn [] @token->index-atom)
+        exit-chan (async/promise-chan)
+        go-chan
+        (async/go-loop [token->index @token->index-atom]
+          (reset! token->index-atom token->index)
+          (let [[msg chan] (async/alts! [body exit-chan] :priority true)
+                next-token->index
+                (condp = chan
+                  body
+                  (when (some? msg)
+                    (let [{:strs [object type]} (try-parse-json msg)]
+                      (case type
+                        "INITIAL"
+                        (reduce
+                          (fn [token->index index]
+                            (assoc token->index (get index "token") index))
+                          {}
+                          object)
+
+                        "EVENTS"
+                        (reduce
+                          (fn [token->index {:strs [object type]}]
+                            (case type
+                              "UPDATE"
+                              (assoc token->index (get object "token") object)
+                              "DELETE"
+                              (dissoc token->index (get object "token") object)
+                              (throw (ex-info "Unknown event type received in EVENTS object" {:event object}))))
+                          token->index
+                          object)
+                        (throw (ex-info "Unknown event type received from watch" {:event msg})))))
+
+                  exit-chan
+                  (do
+                    (async/close! body)
+                    false))]
+            (when next-token->index
+              (recur next-token->index))))]
+    {:exit-chan exit-chan
+     :error-chan error
+     :go-chan go-chan
+     :query-state-fn query-state-fn}))
+
+(defn- start-watches
+  [router-urls cookies]
+  (mapv
+    #(start-watch % cookies)
+    router-urls))
+
+(defn- stop-watch
+  [{:keys [exit-chan go-chan]}]
+  (async/close! exit-chan)
+  (async/<!! go-chan))
+
+(defn- stop-watches
+  [watches]
+  (doseq [watch watches]
+    (stop-watch watch)))
+
+(defmacro assert-watch-token-index-entry
+  [watch token entry]
+  `(let [watch# ~watch
+         token# ~token
+         entry# ~entry
+         query-state-fn# (:query-state-fn watch#)]
+     (is (wait-for
+           #(= (get (query-state-fn#) token#)
+               entry#)
+           :interval 1 :timeout 5))))
```

Comment:
`is` accepts a string as the second argument, we should use it to explain which assertion failed and on which router.